contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+  * Because Flex drivers exist and depend on the Flex interface, it will continue to be supported with a stable API.
+  * The CSI Volume plugin will co-exist with Flex volume plugin.
+
+## Design Overview
+
+To support CSI Compliant Volume plugins, a new in-tree CSI Volume plugin will be introduced in Kubernetes. This new volume plugin will be the mechanism by which Kubernetes users (application developers and cluster admins) interact with external CSI volume drivers.
+
+The `SetUp`/`TearDown` calls for the new in-tree CSI volume plugin will directly invoke `NodePublishVolume` and `NodeUnpublishVolume` CSI RPCs through a unix domain socket on the node machine.
+
+Provision/delete and attach/detach must be handled by some external component that monitors the Kubernetes API on behalf of a CSI volume driver and invokes the appropriate CSI RPCs against it.
+
+To simplify integration, the Kubernetes team will offer a container that captures all the Kubernetes specific logic and acts as a bridge between third-party containerized CSI volume driver and Kubernetes (each deployment of a volume plugin would have it’s own “bridge”).
+
+## Design Details
+
+### Third-Party CSI Volume Drivers
+
+Kubernetes is as minimally prescriptive on the packaging and deployment of a CSI Volume Driver as possible. Use of the *Communication Channels* (documented below) is the only requirement for enabling an arbitrary external CSI compatible storage driver in Kubernetes.
+
+This document recommends a standard mechanism for deploying an arbitrary containerized CSI driver on Kubernetes to simplify deployments of containerized CSI compatible volume drivers on Kubernetes (see the “Recommended Mechanism for Deploying CSI Drivers on Kubernetes” section below), however, this is strictly optional.
+
+### Communication Channels
+
+#### Kubelet to CSI Driver Communication
+
+Kubelet (responsible for mount and unmount) will communicate with an external “CSI volume driver” running on the same host machine (whether containerized or not) via a Unix Domain Socket.
+
+The Unix Domain Socket will be registered with kubelet using the [Device Plugin Unix Domain Socket Registration](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#unix-socket) mechanism. This mechanism will need to be extended to support registration of both CSI volume drivers and device plugins independently.
+
+Upon initialization of the external “CSI volume driver”, some external component should call `GetNodeId` to get the mapping from Kubernetes NodeId to CSI driver NodeId. It should then add the CSI driver NodeId as a label to the Kubernetes Node API object. The key of the label should be `csi.volume.kubernetes.io/<CsiDriverName>/nodeID`. This will enable the component that will issue `ControllerPublishVolume` calls to use the label as a mapping from cluster node ID to storage node ID.
```

Comment:
Does "kubernetes node id" mean node name or node uid or something else?

Reply:
The specified labels don't seem to fall within allowed label keys - https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

Reply:
* fixed "kubernetes NodeId" -> "kubernetes node name"
* fixed `csi.volume.kubernetes.io/<CsiDriverName>/nodeID` -> `nodeid.csi.volume.kubernetes.io/<CsiDriverName>`
* moved from labels to annotations
  * 63 character limit for label values looks too short
  * we might need to put json there, CSI currently describes NodeID as a *dictionary*
  * we don't need to select nodes based on this label

Reply:
* is CsiDriverName expected to contain `/` characters? if so, same issue still exists
* what is the length limit on CsiDriverName?

Reply:
CSI does not say anything about name. I will add a note that the name will be sanitized and even rejected if it's too long / too ugly.